core/services/gateway/handlers/functions/subscriptions.go

```diff
@@ -0,0 +1,207 @@
+		}
+		if s.config.UpdateRangeSize == 0 {
+			return errors.New("OnchainSubscriptionsConfig.UpdateRangeSize must be greater than 0")
+		}
+
+		s.closeWait.Add(1)
+		go s.queryLoop()
+
+		return nil
+	})
+}
+
+func (s *onchainSubscriptions) Close() error {
+	return s.StopOnce("OnchainSubscriptions", func() (err error) {
+		s.lggr.Info("closing onchain subscriptions")
+		close(s.stopCh)
+		s.closeWait.Wait()
+		return nil
+	})
+}
+
+func (s *onchainSubscriptions) GetMaxUserBalance(user common.Address) (*big.Int, error) {
+	s.rwMutex.RLock()
+	defer s.rwMutex.RUnlock()
+	return s.subscriptions.GetMaxUserBalance(user)
+}
+
+func (s *onchainSubscriptions) queryLoop() {
+	defer s.closeWait.Done()
+
+	ticker := time.NewTicker(time.Duration(s.config.UpdateFrequencySec) * time.Second)
+	defer ticker.Stop()
+
+	start := uint64(1)
+	lastKnownCount := uint64(0)
+
+	queryFunc := func() {
+		ctx, cancel := utils.ContextFromChanWithTimeout(s.stopCh, time.Duration(s.config.UpdateTimeoutSec)*time.Second)
+		defer cancel()
+
+		latestBlockHeight, err := s.client.LatestBlockHeight(ctx)
+		if err != nil || latestBlockHeight == nil {
+			s.lggr.Errorw("Error calling LatestBlockHeight", "err", err, "latestBlockHeight", latestBlockHeight.Int64())
+			return
+		}
+
+		blockNumber := big.NewInt(0).Sub(latestBlockHeight, s.blockConfirmations)
+
+		updateLastKnownCount := func() {
+			count, err := s.getSubscriptionsCount(ctx, blockNumber)
+			if err != nil {
+				s.lggr.Errorw("Error getting subscriptions count", "err", err)
+				return
+			}
+			lastKnownCount = count
+		}
+
+		if lastKnownCount == 0 {
+			updateLastKnownCount()
+		}
+		if lastKnownCount == 0 {
+			s.lggr.Info("Router has no subscriptions yet")
+			return
+		}
+
+		if start > lastKnownCount {
+			start = 1
+		}
+
+		end := start + uint64(s.config.UpdateRangeSize)
+		if end > lastKnownCount {
+			updateLastKnownCount()
+			if end > lastKnownCount {
+				end = lastKnownCount
+			}
+		}
+		if err := s.querySubscriptionsRange(ctx, blockNumber, start, end); err != nil {
+			s.lggr.Errorw("Error querying subscriptions", "err", err, "start", start, "end", end)
+			return
+		}
+
+		start = end + 1
+	}
+
+	queryFunc()
+
+	for {
+		select {
+		case <-s.stopCh:
+			return
+		case <-ticker.C:
+			queryFunc()
+		}
+	}
+}
+
+func (s *onchainSubscriptions) querySubscriptionsRange(ctx context.Context, blockNumber *big.Int, start, end uint64) error {
+	subscriptions, err := s.router.GetSubscriptionsInRange(&bind.CallOpts{
+		Pending:     false,
+		BlockNumber: blockNumber,
+		Context:     ctx,
+	}, start, end)
+	if err != nil {
+		return errors.Wrap(err, "unexpected error during functions_router.GetSubscriptionsInRange")
+	}
+
+	s.rwMutex.Lock()
+	defer s.rwMutex.Unlock()
+	for i, subscription := range subscriptions {
+		subscriptionId := start + uint64(i)
+		subscription := subscription
+		s.subscriptions.UpdateSubscription(subscriptionId, &subscription)
+	}
+
```

Comment:
Add log messages with number of subs updated.

Reply:
This will always be equal to `UpdateRangeSize` except for the tail. Not sure this is useful at all.

Reply:
you're right